### PR TITLE
Checkout: Add placeholder option to country and state select, add error validation

### PR DESCRIPTION
**Pull request by samueljseay**
After migrating the country and state selection in checkout to a native select recently we uncovered some issues / undefined behaviours that arise due to the different input being used.

This addresses a couple of problems:

1. When the store is set to not choose a default shipping region, we don't want to auto-select the top country option in the select as it often won't make sense, e.g. defaulting to Afghanistan due to alphabetical order will not make sense in most cases. To solve this we introduce an empty placeholder option. It's selected by default in this case but also disabled so it cannot be re-selected

2. Now we have an empty placeholder option it's easy to submit the form without a selected country or state. To solve this, I've introduced client-side validation for country and state when required. I've used existing validation components to keep the styling consistent.

### How to test

0. You'll need a basic store setup with a few in stock products, ensure that this setting is set to "No location by default":

image

#### Scenario 1: When "No location by default" is enabled the country select will have a disabled placeholder selected

1. Add a product to cart and proceed from cart to checkout
2. In Checkout see that the country select has a placeholder value selected by default:

Screenshot 2024-07-19 at 3 01 51 PM
Screenshot 2024-07-19 at 3 01 55 PM

#### Scenario 2: Submitting "Place order" will validate country is missing

1. From the state of Scenario 1 enter any text in the "City" field. 
2. A validation error will show on the country select:

Screenshot 2024-07-19 at 3 06 23 PM

#### Scenario 3: Selecting a country will remove the validation error

1. From the state of  Scenario 2 click on the country select and select a country, the country validation error should disappear:

Screenshot 2024-07-19 at 3 07 41 PM

#### Scenario 4: Select a country with selectable "state" should have an empty option placeholder.

1. From the state of Scenario 3, change the country select to United States
2. The State select should now have a placeholder option in it:

Screenshot 2024-07-19 at 5 33 28 PM

#### Scenario 5: Trying to submit with a selected country where state is required (but not selected) should show a validation error.

1. With United States selected as in Scenario 4, click "place order"
2. The state select should now show a validation error:

Screenshot 2024-07-19 at 5 34 49 PM

#### Scenario 6: Changing from a country with required state to a country with optional state should not show error

1. With the validation error showing for empty state with United States, change the country selector to Greece
2. The state select should return to a non error state:

Screenshot 2024-07-19 at 5 20 27 PM

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message
Add placeholder options and validation to checkout country and state select inputs.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	values: ShippingAddress,
+	isRequired: boolean
+) => {
+	const validationErrorId = `${ addressType }_state`;
+	const hasValidationError =
+		select( VALIDATION_STORE_KEY ).getValidationError( validationErrorId );
+
+	const countryChanged = lastValue( values.country ) !== values.country;
+
+	if ( hasValidationError ) {
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts:46]: A semi-complex set of states to determine if we should show a validation error for the state.

If we choose a country that has a required state then we "set" the validation error (which puts it hidden in state). If the user tries to submit with the same country we show the error.

If the country is changed to a non-required state we remove the validation, if its changed to another country with required state the validation is set but hidden.

File: plugins/woocommerce-blocks/assets/js/base/components/country-input/billing-country-input.tsx [hunk 2]
```diff
@@ -10,8 +10,7 @@ import CountryInput from './country-input';
 import type { CountryInputProps } from './CountryInputProps';
 
 const BillingCountryInput = ( props: CountryInputProps ): JSX.Element => {
-	//  TODO - are errorMessage and errorId still relevant when select always has a value?
-	const { errorMessage: _, errorId: __, ...restOfProps } = props;
+	const { ...restOfProps } = props;
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/assets/js/base/components/country-input/billing-country-input.tsx:13]: We now pass the errorMessage and id through to perform validation in the CountryInput

File: plugins/woocommerce-blocks/assets/js/base/components/state-input/state-input.tsx [hunk 3]
```diff
@@ -34,18 +41,37 @@ const StateInput = ( {
...
-			countryStates
-				? Object.keys( countryStates ).map( ( key ) => ( {
-						value: key,
-						label: decodeEntities( countryStates[ key ] ),
-				  } ) )
-				: [],
-		[ countryStates ]
-	);
+	const options = useMemo< SelectOption[] >( () => {
+		if ( countryStates && Object.keys( countryStates ).length > 0 ) {
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/assets/js/base/components/state-input/state-input.tsx:48]: In some cases countryStates is an empty array causing us to add a select with an empty option (before this didn't happen because we didn't have a empty placeholder option). This fixes that case.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
+/**
+ * External dependencies
+ */
+import { dispatch, select } from '@wordpress/data';
+import { ShippingAddress } from '@woocommerce/settings';
+import { VALIDATION_STORE_KEY } from '@woocommerce/block-data';
+import { __ } from '@wordpress/i18n';
+
+function previousString( initialValue?: string ) {
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: Since there are no hooks in the other validation I wanted to keep the same approach here. This simple function keeps track of the previous value because we need to know if the country changed when determining to show validation of the state.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx [hunk 5]
```diff
@@ -56,6 +62,11 @@ export const Select = ( props: SelectProps ) => {
 							key={ option.value }
 							value={ option.value }
 							data-alternate-values={ `[${ option.label }]` }
+							disabled={
```
Comment (samueljseay) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:65]: Added support to have disabled options. I think its simpler and more flexible to allow passing disabled options than to have the select have a concept of a "placeholder" option that would complexify the API of this component (imo) needlessly.

**Comment by nielslange** [comment]
@samueljseay I retriggered the CI pipeline by closing and reopening this PR.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 4 above)
Comment (opr) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: Makes sense, but I was going to suggest turning them _into_ hooks, that way we can make use of the existing [`usePrevious`](https://github.com/woocommerce/woocommerce/blob/37d482368029ad1f2e0d85669648ae69652a5e23/plugins/woocommerce-blocks/assets/js/base/hooks/use-previous.ts#L16) hook rather than making the `previousString` function, and we get to use `useSelect` too. I believe this comes with some performance benefit.

Not sure if it makes sense to do the change in this PR, given `validateCountry` isn't one, but just thought I'd give the suggestion, maybe we can tackle it in a follow up.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+	const countryChanged = lastValue( values.country ) !== values.country;
+
+	if ( hasValidationError ) {
+		if ( ! isRequired || values.state ) {
+			// Validation error has been set, but its no longer required, or the state was provided, clear the error.
+			dispatch( VALIDATION_STORE_KEY ).clearValidationError(
+				validationErrorId
+			);
+		} else if ( ! countryChanged ) {
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: I found this and the comment below a little hard to follow, would it be better to check if `values.country === ''` just to make it easier to reason with?

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+) => {
+	const validationErrorId = `${ addressType }_state`;
+	const hasValidationError =
+		select( VALIDATION_STORE_KEY ).getValidationError( validationErrorId );
+
+	const countryChanged = lastValue( values.country ) !== values.country;
+
+	if ( hasValidationError ) {
+		if ( ! isRequired || values.state ) {
+			// Validation error has been set, but its no longer required, or the state was provided, clear the error.
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: Minor grammar nitpick, sorry 😖 
```suggestion
			// Validation error has been set, but it's no longer required, or the state was provided, clear the error.
```

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx [hunk 8]
```diff
@@ -56,6 +62,11 @@ export const Select = ( props: SelectProps ) => {
...
+								option.disabled !== undefined
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:66]:
```suggestion
								typeof option.disabled !== 'undefined'
```

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx [hunk 9]
```diff
@@ -9,11 +9,17 @@ import { useCallback, useId } from '@wordpress/element';
...
+	label: string;
+	disabled?: boolean;
+};
+
 type SelectProps = Omit<
 	React.SelectHTMLAttributes< HTMLSelectElement >,
 	'onChange'
 > & {
-	options: { value: string; label: string }[];
+	options: { value: string; label: string; disabled?: boolean }[];
```
Comment (opr) [inline plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: can we change this to `SelectOption[]`?

```suggestion
	options: SelectOption[];
```

**Review by opr — CHANGES_REQUESTED** [review]
Looks great @samueljseay thanks for this, it's testing well too in different browsers

I'd love to see some tests for this as well, do you plan to add them in a followup?

Just left a couple of minor comments/suggestions. I'm interested to hear your thought on changing the validation functions to hooks as well, even if we don't do it in this PR is it something worth looking into?

**Review by nielslange — COMMENTED** [review]
Thanks for working on this PR, @samueljseay. I've successfully tested the use cases you mentioned. I see that @opr left some feedback and @senadir discovered that when the state select field contains an error, that the scroll jumps to the wrong section. He mentioned this in p1721654388438629-slack-C8X6Q7XQU. Let's make sure we address @opr's and @senadir's points before merging this PR. Please feel free to request another review from me once these points have been addressed.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/form.tsx [hunk 10]
```diff
@@ -93,15 +94,22 @@ const Form = < T extends AddressFormValues | ContactFormValues >( {
 		}
 	}, [ onChange, addressFormFields, values ] );
 
-	// Maybe validate country when other fields change so user is notified that it's required.
+	// Maybe validate country and state when other fields change so user is notified that they're required.
 	useEffect( () => {
-		if (
-			addressType === 'shipping' &&
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/form.tsx:99]: Not clear to me why we no longer limit this to shipping, or why we limited it to shipping in the first place.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/form.tsx [hunk 11]
```diff
@@ -93,15 +94,22 @@ const Form = < T extends AddressFormValues | ContactFormValues >( {
...
-			objectHasProp( values, 'country' )
-		) {
-			validateShippingCountry( values );
+		if ( objectHasProp( values, 'country' ) ) {
+			validateCountry( addressType, values );
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/form.tsx:100]: Ah I see we moved this to `validateCountry`

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 1 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts:46]: I'm wondering why we need to keep track of previous countries and not simply check the current state value and requirement? Is this because unsetting can be done via user (show error) or via country change (hide error)?

File: plugins/woocommerce-blocks/assets/js/base/components/country-input/country-input.tsx [hunk 12]
```diff
@@ -4,13 +4,23 @@
...
  * Internal dependencies
  */
 import './style.scss';
 import type { CountryInputWithCountriesProps } from './CountryInputProps';
-import { Select } from '../select';
+import { Select, SelectOption } from '../select';
+
+const emptyCountryOption: SelectOption = {
+	value: '',
+	label: __( 'Select a country', 'woocommerce' ),
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/country-input/country-input.tsx]: I assume this is the default for this one? Ideally we should just use `Select a ${label}` which is used in other places.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+	} else if (
+		! hasValidationError &&
+		isRequired &&
+		! values.state &&
+		values.country
+	) {
+		// No validation has been set yet, if it's required, there is a country set and no state, set the error.
+		dispatch( VALIDATION_STORE_KEY ).setValidationErrors( {
+			[ validationErrorId ]: {
+				message: __( 'Please select your state', 'woocommerce' ),
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: error should `Please select a ${stateField.label}`

**Review by senadir — APPROVED** [review]
This is testing fine with me except the issue with focusing back on errors which I reported in slack.

Otherwise I left some small comments, but pre-approving.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 4 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: @opr I think there is a bit of a challenge introducing hooks here because we conditionally run the validation logic based on `objectHasProp( values, 'country' )` which breaks the rules of hooks.

Could we consider a follow-up to do this, because i don't think it adds any real benefit here. Knowing the implementation of `usePrevious` i can tell you I don't think it has any performance benefit in this situation. We store one string at all times and we only run this code when both the `objectHasProp` condition fulfills and when the `useEffect` deps change. Of course doing things the "hooks" way in general is more consistent but I think that it would make sense to refactor both validations in that case and a separate PR might be better to isolate and test those changes.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 1 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts:46]: > Is this because unsetting can be done via user (show error) or via country change (hide error)?
@senadir Yes exactly.

I needed to distinguish the valid situation where the country was changed to one with an optional state to hide the error.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 7 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: Thanks!

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 6 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: The comment should say "there has not been a country change" that distinguishing factor is important here. I will fix the comment, but it's important not to show the error if the country changed because we need to reset the validation in that case to determine if the new state is required or not.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 8 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:66]: @opr whats your reason for this change? There is not a situation here where option would be undefined so it's quite safe in it's current iteration. we could also do `option?.undefined` to make it slightly more correct I suppose.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 9 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: Good spotting, thanks I'll fix this.

Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx]: Resolved in [9a36764](https://github.com/woocommerce/woocommerce/pull/49616/commits/9a367642a29a1d5b7f3c74f39469402bf4fa2ea6)

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 13 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: @senadir should it be the same for the country validation too then?

https://github.com/woocommerce/woocommerce/blob/9a367642a29a1d5b7f3c74f39469402bf4fa2ea6/plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-country.ts#L28-L34

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/form.tsx (hunk 10 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/form.tsx:99]: @senadir So I'm not 100% on this, but I assume the logic was that billing address is by default pre-filled from the shipping address, but this does not cover the edge case where a user unchecks `Use same address for billing` before every entering a country. In my testing if you do that it and then start filling other fields of the billing address it feels totally reasonable to validate in the same way as shipping. It feels more consistent this way even if the user is unlikely to take that path.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 6 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: @opr I've adjusted this logic because I found one case where it didn't work as expected when you change the address and it shouldn't show validation for the state yet. I've adjusted the comment to better match as well

[adjust validation to handle any address change, not just country. tra…](https://github.com/woocommerce/woocommerce/pull/49616/commits/662657168ad158e0ecde07e07dfd14de6f30214c) 
6626571

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 13 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: I've fixed the state issue, I don't mind fixing the country here too, but could also be done as a follow up since this behaviour was already there.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 1 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts:46]: I've adjusted the logic of this because we also need to determine if other address fields change. If they changed we don't want to show the state validation (ie if you edit "City", we don't need to immediately inform you that state is required, it doesn't feel necessary). We only want to show it if the user tried to submit the form without it.

The only thing we don't have is an "on blur" validation, but in testing (imo) I don't think this impacted ux. Of course this is all just opinion and I'm open to changing that :)

File: plugins/woocommerce-blocks/assets/js/base/components/country-input/country-input.tsx (hunk 12 above)
Comment (samueljseay) [reply plugins/woocommerce-blocks/assets/js/base/components/country-input/country-input.tsx]: This is fixed now, thanks for spotting this.

**Comment by samueljseay** [comment]
Hey @opr @nielslange @senadir would you mind taking another look.

And specifically to address a couple of points from @opr :

 1. I agree this should have some tests, I'm just having some technical issues with the existing JS unit tests, it looks like the previously rendered components are leaking between tests. I am going to see if @nielslange can get [his PR](https://github.com/woocommerce/woocommerce/pull/48843) merged first because it will have significant impact on how the tests are rendered/updated.  I have created a follow up task for that: https://github.com/woocommerce/woocommerce/issues/49896
 
 2. The conversion of validateState/validateCountry to hooks can certainly be done but I'd also like to propose prioritizing that separately. I've created an issue for that https://github.com/woocommerce/woocommerce/issues/49897 and my justification for not doing it here is discussed in [this comment](https://github.com/woocommerce/woocommerce/pull/49616#discussion_r1689011770)

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts (hunk 4 above)
Comment (opr) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/validate-state.ts]: Agree with your points, but I wasn't suggesting this with a performance benefit in mind, rather making it easier to reason with the code, but I am fine with deferring this until later.

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 8 above)
Comment (opr) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:66]: I feel my suggestion is more robust and handles the case where `option` is undefined, however unlikely that is. I don't have super strong feelings about it, so I'm fine for you to ignore this suggestion.

**Review by opr — APPROVED** [review]
This is working for me, thanks for addressing the code changes I suggested. From a UX perspective, the changes to the state validation feel better.

I don't feel strongly about the [`typeof` change suggested](https://github.com/woocommerce/woocommerce/pull/49616/files#r1686451598) so don't mind if this is merged with it unaddressed.

The failing CI job is failing because it can't upload an artifact. The tests themselves passed.

**Comment by senadir** [comment]
Let's merge and continue, as this is critical for 9.2

File: plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx (hunk 5 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/base/components/select/index.tsx:65]: Turns out we did need to support this at component level because additional select inputs are also broken
